shared/codec/codec.go

```diff
@@ -5,56 +5,45 @@ import (
 	"google.golang.org/protobuf/types/known/anypb"
 )
 
-// TODO: Use generics in place of `proto.Message` in the interface below
-//       so every caller does not need to do in place casting.
-type Codec interface { // TODO (Team) move to shared. Possibly rename
+// CONSIDERATION: Use generics in place of `proto.Message` in the interface below so
+//                every caller does not need to do in place casting.
+type Codec interface {
 	Marshal(proto.Message) ([]byte, error)
 	Unmarshal([]byte, proto.Message) error
 	ToAny(proto.Message) (*anypb.Any, error)
 	FromAny(*anypb.Any) (proto.Message, error)
+	Clone(proto.Message) proto.Message
 }
 
-var _ Codec = &ProtoCodec{}
+type ICodec proto.Message
```

Comment:
I don't like this TBH. Care to elaborate the reasoning behind the change here?

Reply:
As far as I understand you are trying to centralize the type that we can encode/decode. Right?

In that case I guess all the methods should use that and not `proto.Message`. Something like: 

```go
package codec

import (
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
)

// DOCUMENT: Some of the types through the code based (e.g. Transaction) are defined via serializable
// files (e.g. .proto files) and add functionality defined in go on top of it (e.g. ITransaction). This
// a very implicit design pattern that may be non-obvious to new developers.
type CodecType proto.Message

// CONSIDERATION: Use generics in place of `proto.Message` in the interface below so
//
//	every caller does not need to do in place casting.
type Codec interface {
	Marshal(CodecType) ([]byte, error)
	Unmarshal([]byte, CodecType) error
	ToAny(CodecType) (*anypb.Any, error)
	FromAny(*anypb.Any) (CodecType, error)
	Clone(CodecType) CodecType
}

var _ Codec = &protoCodec{}

// IMPROVE: Need to define a type similar to `type ProtoAny anypb.Any` so that we are
//
//	referencing protobuf specific types (e.g. anypb.Any) anywhere in the codebase.
type protoCodec struct{}

func GetCodec() Codec {
	return &protoCodec{}
}

// IMPROVE: If/when we move Pocket's `Error` type into a separate package, we can return `ErrProtoMarshal` here
func (p *protoCodec) Marshal(msg CodecType) ([]byte, error) {
	return proto.MarshalOptions{Deterministic: true}.Marshal(msg)
}

func (p *protoCodec) Unmarshal(bz []byte, msg CodecType) error {
	return proto.Unmarshal(bz, msg)
}

func (p *protoCodec) ToAny(msg CodecType) (*anypb.Any, error) {
	return anypb.New(msg)
}

func (p *protoCodec) FromAny(any *anypb.Any) (CodecType, error) {
	return anypb.UnmarshalNew(any, proto.UnmarshalOptions{})
}

func (p *protoCodec) Clone(msg CodecType) CodecType {
	return proto.Clone(msg)
}
```
but at that point I would take care of `FromAny` and `ToAny` since are protobuf specific.

Also, question: are we planning to support different encodings apart from proto? If not, this sounds a bit overkill and prematurely optimized. I'd just restore the `proto.Message` embedding in the `Message`

Thoughts?